aiida_optimize/engines/_particle_swarm.py

```diff
@@ -0,0 +1,268 @@
+
+        self.particles = np.array(particles)
+        n_vars = len(self.particles[0])
+        assert all([len(part) == n_vars for part in self.particles])
+        self.velocities = velocities
+
+        self.global_best = global_best
+        self.fun_global_best = fun_global_best
+        self.local_best = local_best
+        self.fun_local_best = fun_local_best
+
+        self.max_iter = max_iter
+        self.num_iter = num_iter
+
+        self.input_key = input_key
+        self.result_key = result_key
+
+        self.next_submit = next_submit
+        self.next_update = next_update
+
+        self.finished = finished
+        self.exceeded_max_iters = exceeded_max_iters
+
+    @submit_method(next_update='update_general')
+    def submit_initialize(self):
+        n_parts = len(self.particles)
+        n_vars = len(self.particles[0])
+        self.local_best = self.particles
+        self.fun_local_best = np.full(n_parts, np.inf)
+        self.fun_global_best = np.inf
+        self.velocities = np.zeros((n_parts, n_vars))
+        self._logger.report('Submitting first step.')
+        return [self._to_input_list(x) for x in self.particles]
+
+    def _to_input_list(self, x):
+        input_list = orm.List()
+        input_list.extend(x)
+        return {self.input_key: input_list}
+
+    @update_method(next_submit='new_iter')
+    def update_general(self, outputs):  # pylint: disable=missing-function-docstring
+        fun_particles = np.array(self._get_values(outputs))
+        for index, val in enumerate(fun_particles):
+            if val < self.fun_local_best[index]:
+                self.fun_local_best[index] = val
+                self.local_best[index] = self.particles[index]
+        for index, val in enumerate(self.fun_local_best):
+            if val < self.fun_global_best:
+                self.fun_global_best = val
+                self.global_best = self.local_best[index]
+
+    def _get_values(self, outputs):
+        return [get_nested_result(res, self.result_key).value for _, res in sorted(outputs.items())]
+
+    def create_particle(self):  # pylint: disable=missing-function-docstring
+        n_var = len(self.particles[0])
+        new_vel = deepcopy(self.velocities)
+        for idx, val in enumerate(self.particles):
+            new_vel[idx] = [
+                self.update_vel(
+                    OMEGA, self.velocities[idx][i], C1, C2, val[i], self.local_best[idx][i],
+                    self.global_best[i]
+                ) for i in range(n_var)
+            ]
+
+        new_parts = deepcopy(self.particles)
+        for fd in range(len(self.particles)):
+            new_parts[fd] = np.array(new_vel[fd]) + self.particles[fd]
+
+        return np.array(new_parts), np.array(new_vel)
+
+    @staticmethod
+    def update_vel(omega, v, c1, c2, x, pi, pg):  # pylint: disable=too-many-arguments
+        return omega * v + c1 * uniform(0, 1) * (pi - x) + c2 * uniform(0, 1) * (pg - x)
```

Comment:
I wonder if we need to add some logic to ensure we get "good" randomness behavior here. I'm more familiar with Monte Carlo type calculations, but there one would typically seed the RNG with a known number, to get reproducible results.

Here, the trouble is that each step can happen from a new / random Python interpreter. This presents a challenge because we can't simply rely on the RNG remaining the same. So even if we don't care about reproducibility, it could mean that we don't get "good" randomness.
I suspect a solution to this would include either persisting the RNG state between steps (maybe `numpy.random` can help with this?), or re-seeding and then re-iterating through the previous values (keeping track of how many random numbers we used).

Reply:
After discussion with @vdikan, we implemented to pass the state of the RNG among steps, using `get_state` and `set_state` of `numpy.random`. Let us know what you think.

Reply:
Yep, that looks good to me! :+1: